UX Overhaul Part 1

therobbiedavis opened the pull request:
![image](https://github.com/user-attachments/assets/ff5c371e-dc52-498b-99c8-ac083a72a295)

![background animation - phase 4 - completion](https://github.com/user-attachments/assets/9b5ecdb7-995b-436e-a560-66cff8d5417b)

![image](https://github.com/user-attachments/assets/5f3d4855-67ef-4480-9b91-076866754ccd)

![image](https://github.com/user-attachments/assets/93942ac3-04e3-4b02-87e2-1367e1d84929)


### This is the first pass and initial PR to bring the new UX to Kavita. Note that Series detail and all screens around that (outside of the colorscape functionality) will follow in a separate PR due to the size of the changes. Nightly testers please leave copious amounts of feedback. Our ears are open. 

# Added
- Added: Replaced default Dark theme which will begin our UX overhaul. The previous dark theme is still available on the theme repo.
- Added: Added a lot of ability to theme the login screen, including the background image. 
- Added: Added ColorScapes. ColorScapes is a new dynamic tinting system for Kavita. During cover generation, Kavita will pick primary colors from the image and apply them to the background on pages where the image is a primary element (series detail for now). These are configurable via the theming system. (Light themes do not look good with this system, it should be disabled)
- Added: Added a new actionable for Library to generate colorscapes (which is a soft refresh covers)

# Changed
- Changed: Changed the Login look and feel.
- Changed: Changed the max password character length to 256.
- Changed: Changed some headers, spacing, and positioning to make things more uniform across the app.
- Changed: Revamped how settings/preferences work in the app. Following Plex' style, they are all on one page (RBS applies) with a dedicated left nav menu that groups based on functionality. (Sorry localizers, lots of string updates)
- Changed: When navigating away from library/collection/bookmark/etc-detail, instead of just restoring position when interacting with a jumpkey, Kavita will now restore position down to the location of jump point in all cases.
- Changed: Jump keys now show an animation instead of a tooltip to show number of series
- Changed: Filter icon now shows an animation when actively configuring it
- Changed: Moved some screens around to dedicated sections and most of Kavita+ screens are now on a new Scrobbling page (this is temp and will be updated with #2979)
- Changed: Moved the customize modal into a dedicated settings item to make discovery of the feature easier. 
- Changed: Lots of small consistency issues and QoL tweaks on different pages to bring a better UX and interaction. A lot of functionality has been moved to drive a consistent experience and discovery of features. 

# Fixed
- Fixed: Fixed an issue where autofilled passwords were getting truncated due to the max character length.
- Fixed: Fixed a side-nav css var that was inappropriately assigned. (Fixes #3043)
- Fixed: Fixed a bug in theme manager where the active pill wouldn't have the correct color for non-dark themes
- Fixed: Fixed a bug where reading list cards wouldn't promote/unpromote from actionables
- Fixed: Fixed a bug where OPDS reading list feeds were missing pagination
- Fixed: Fixed an edge case where busting Kavita+ cache in edit series then hitting close wasn't refreshing screen.

# Theme
- There were a ton of changes and new css variables. Wiki will be updated and extra documentation around colorscape system. 
- Tables are now themeable

Closes #3052

therobbiedavis committed:
login flow uodates

therobbiedavis committed:
Merge branch 'develop' into feature/revamped-ux

therobbiedavis committed:
style updates and increase to password max length

therobbiedavis committed:
Merge branch 'develop' into feature/revamped-ux

therobbiedavis committed:
Updating default theme and login ux

therobbiedavis commented:
@majora2007 I think there may need a migration in order to either update or remove the previous dark theme from the user's databases since I moved that over to the theme repo.

majora2007 commented on UI/Web/src/app/registration/_components/register/register.component.scss:
```diff
@@ -1,4 +1,35 @@
+    }
+
+    &::placeholder {
+        opacity: 0.5;
```
We probably want to put this in a css variable as well.

majora2007 commented on UI/Web/src/app/registration/_components/confirm-email-change/confirm-email-change.component.scss:
```diff
@@ -3,5 +3,9 @@
+}
+
+p {
+    font-family: "League Spartan", sans-serif;
```
Shouldn't this be: `font-family: var(--login-input-font-family);`

majora2007 commented on UI/Web/src/app/registration/_components/splash-container/splash-container.component.html:
```diff
@@ -1,12 +1,26 @@
+            <div class="row align-items-center row-cols-1 justify-content-center">
+                <div class="col col-lg-8 col-md-10 col-sm-10 col-xs-10 p-0 position-relative">
+                    <div class="justify-content-center">
+                        <div class="logo"></div>
```
Place aria-hidden="true" on this since it's just the logo.

majora2007 commented on UI/Web/src/app/registration/_components/splash-container/splash-container.component.html:
```diff
@@ -1,12 +1,26 @@
             <span>
                 <div class="logo-container">
-                    <h3 class="card-title text-center">
+                    <div class="card-title text-center">
```
If we can still use an h3 and override the styles that would be great.

therobbiedavis replied:
It doesn't make sense to leave it as an h3 because the title is pulled into an h2, so you're wrapping an h2 with an h3.
![image](https://github.com/Kareadita/Kavita/assets/7907265/ec2ddaad-f13d-45a7-b694-953d30751555)

majora2007 replied:
Ahh nice catch

majora2007 commented on UI/Web/src/app/registration/user-login/user-login.component.scss:
```diff
@@ -11,11 +11,38 @@ a {
+
+    &::placeholder {
+        opacity: 0.5;
+        color: #cecece;
```
Let's css variable this as well.

majora2007 reviewed (changes requested):
A few comments overall. As requested in discord, let's revert the changes to the system them and move the current Dark theme to Theme repo as Dark-Legacy given this is the start of the new UX.

therobbiedavis committed:
Lots of style fixes and uniformity fixes

therobbiedavis committed:
PR Review feedback

therobbiedavis committed:
Revert change

majora2007 commented on UI/Web/src/app/_services/nav.service.ts:
```diff
@@ -79,9 +79,9 @@ export class NavService {
-    this.renderer.setStyle(this.document.querySelector('body'), 'margin-top', '56px');
-    this.renderer.setStyle(this.document.querySelector('body'), 'height', 'calc(var(--vh)*100 - 56px)');
-    this.renderer.setStyle(this.document.querySelector('html'), 'height', 'calc(var(--vh)*100 - 56px)');
+    this.renderer.setStyle(this.document.querySelector('body'), 'margin-top', '75px');
```
Okay this number change has big impacts. We probably want to make it a css variable. Book reader also uses it and I think bulk actions component (just search for the number).

therobbiedavis replied:
Changed to css var and added it where appropriate. Tested the book reader and fixed bulk actions.

majora2007 commented on UI/Web/src/app/_single-module/card-actionables/card-actionables.component.scss:
```diff
@@ -26,3 +26,8 @@
     padding: var(--bs-dropdown-item-padding-y) 0;
 }
+
+#actions-card {
```
Please avoid using ids.

majora2007 commented on UI/Web/src/theme/themes/dark.scss:
```diff
@@ -2,6 +2,16 @@
 
 :root, :root .default {
     /* Base colors */
+    --layer0: rgba(255,255,255,0);
```
What are these layers? It's not clear from usage either. 

I'm not sure why we have a material file either?

therobbiedavis replied:
Changed the names to make it more clear and some documentation, but they refer to elevation layers.
https://m2.material.io/design/color/dark-theme.html#properties
![image](https://github.com/Kareadita/Kavita/assets/7907265/47fe102b-c552-4e36-84b0-3d84d89cceba)

therobbiedavis committed:
More feedback updates

- Updated nav-offset to css var
- Change ID selector to class selector
- Fixed more uniformity
- Added better naming to the layers and documentation

therobbiedavis committed:
Fixing bulk component

therobbiedavis committed:
Small fix for e-ink theme

Fixes #3043

majora2007 commented on UI/Web/src/app/registration/_components/splash-container/splash-container.component.html:
```diff
@@ -1,12 +1,26 @@
-<div class="mx-auto login" [ngStyle]="{'height': (navService.navbarVisible$ | async) ? 'calc(var(--vh, 1vh) * 100 - 57px)' : 'calc(var(--vh, 1vh) * 100)'}">
+<div class="mx-auto container login text-center"
```
I see -57px. We should use --nav-offset

therobbiedavis replied:
Fixed.

therobbiedavis committed:
Feedback update 3

- Fixd the splashpage to use the navbar-offset css var

majora2007 commented:
The backend will only allow 32 characters, as you can see in the Dtos. 

Do you mean to update this as well? 
[StringLength(32, MinimumLength = 6)]
    public string Password { get; set; } = default!;

majora2007 committed:
Implemented the ability for Kavita to pick and save primary/secondary color for anything that has a cover image.

majora2007 commented:
As discussed in person, there are a few issues around broken styles on modals and whatnot. I will in addition be implementing the accent functionality in this PR so first release is useable. 

We can consider this PR on hold until we get through these things (with holiday upcoming, expect 1 month).

majora2007 committed:
Rendering of the colored canvas is now working properly (but still hacky).

Need to do a few things:
- Fix the primary color calculation
- Make the frontend respond to cover events and refresh/apply colors.
- Dedicated APIs for color scape stuff.
- Theme-ify it

majora2007 committed:
Fixed a bug where page color wasn't getting unset during page navigations.

majora2007 committed:
Refactored how the background colorscape is applied to use css variables injected rather than inline styles.

majora2007 committed:
Refactored the offset calculation to be tweakable by themes.

majora2007 committed:
Got the ordering working for color selection. Secondary color isn't complementary yet. Lots of alternative code in here, still working it out.

majora2007 committed:
Converted card actionables to control flow

majora2007 committed:
Added some more test cases for validation on backend. Updated the theme code so you can turn off the transition effect with a theme variable explicitly. Working on the transitioning of the effect between pages to not be so jarring.

majora2007 committed:
Added some more test cases for validation on backend.

majora2007 committed:
Fixed a regression on login page with control flow. Ensure that the color scape doesn't work on the login page.

Removed a style from Robbie that broke a lot of styling throughout the app.

majora2007 committed:
Implemented the ability to transition dynamically from one colorscape to another. The transition time is determined by the distance between the colors.

majora2007 committed:
Cleaned up the code and started working on the opacity that Plex has to get the background effect much nicer.

majora2007 committed:
Implemented hooks so that when a cover update occurs, the UI reloads the colorscape for that page.

majora2007 committed:
Fixed some parsing bugs around the default colors no longer being in hex.

majora2007 committed:
Refactored colorscape service to use types as much as possible. If we are transitioning to the same color, we can avoid it and skip the extra work.

majora2007 committed:
Added code to help build light theme colorscapes but after testing, the system doesn't seem to work well. Light theme users should ideally disable.

majora2007 committed:
Removed some debug code. Colorspaces are done. Moving on.

majora2007 committed:
Fixed a bug in theme manager where the active pill wouldn't have the correct color for non-dark themes.

majora2007 committed:
Refactored the user preferences code to its own component to prepare for the upcoming Settings overhaul.

majora2007 committed:
User preferences "dashboard" is now all individual components and ready for upcoming settings overhaul.

majora2007 committed:
Aligned the admin dashboard to use defer tags like user preferences.

majora2007 committed:
Step 1: When on a preference page, the default side nav will not show.

majora2007 committed:
Converted current side nav to control flow syntax

majora2007 committed:
Changed the active side nav item to have transparent background.

Refactored more sidenav code to control flow.

majora2007 committed:
Hooked up the admin section so that it can activate the correct routes.

majora2007 committed:
Refactored side-nav-item to be able to handle fragments.

majora2007 committed:
Added a new preference page to handle all the logic. Refactored the nav header to control flow.

majora2007 committed:
New settings page is in place and functional. Individual components do need style updates, but overall everything is in place.

majora2007 committed:
Removed dead code and updated links to point to the updated urls.

majora2007 committed:
Started redesigning a component for settings page.

Added a nice flip effect for jumpbar.

majora2007 committed:
Made the filter icon a bit more stylistic.

majora2007 committed:
Fixed up the hover area for a jummp key

majora2007 committed:
Control flow card detail layout as much as possible. Still not possible to do the individual cards.

majora2007 committed:
Cleaned up the logic for resuming position on library-detail and the like. It is now much more pinpoint and works on custom sorts.

majora2007 committed:
Some cleanup to the html

majora2007 committed:
Converted Scrobble Tokens to new UX and changed up the edit flow slightly.

majora2007 committed:
Converted Manage Email (need a Robbie css pass)

majora2007 committed:
Account settings page is converted to the new UX (still needs a pass over it though)

majora2007 committed:
Control flow updates

majora2007 committed:
Added a dedicated setting-title component since the design is the same across the board.

majora2007 committed:
Updated User Preferences page to the new UX. Needs accessibility pass, css pass, and needs to be updated to automatically save.

majora2007 committed:
Refactored the save mechanic to just be automatically reflecting.

majora2007 committed:
Refactored all the view modes of user preferences page to use pipes to show localized data.

majora2007 committed:
Updated settings page to dictate which pages are col-5 or not.

majora2007 committed:
Cleaned up the Account settings once again with some extra detail.

majora2007 committed:
Started adding a more customized table implementation for theme and a better match for dark mode.

majora2007 committed:
Indented code

majora2007 committed:
Added height on series detail to avoid it causing layout shift (does not work)

majora2007 committed:
Started working on the button styles to match to the new theme system.

majora2007 committed:
Started working on refactoring device UX.

majora2007 committed:
Device UX is finished until Robbie gives more direction.

majora2007 committed:
Started to refactor where the Kavita+ screens will live (in their own section).

majora2007 committed:
Moved metadata breakdown into Kavita+ page (temp) and introduced another component to help easily find kavitaplus stuff.

majora2007 committed:
All stats components are now stacked on top of each other. Will revisit later to cleanup order and overall UX.

majora2007 committed:
Updated Server settings to the new UX pattern.

majora2007 committed:
Cleaned up some bugs from porting.

majora2007 committed:
Moved all of the scrobbling stuff into one page. This isn't final, just a temp work before the Kavita+ overhaul work (slated for same release as UX overhaul).

majora2007 committed:
Fixed up the OPDS page to stick to the new style. Styling is broken, but Robbie is aware.

majora2007 committed:
Removed all cases of hoverable tables from the app.

majora2007 committed:
Converted Media settings page over to new UX.

majora2007 committed:
Converted Email settings page over to new UX.

majora2007 committed:
Refactored the devices screen to use a table and a modal for the add/edit flow.

majora2007 committed:
Removed table hover class

majora2007 committed:
Updated Tasks screen to new UX.

Now I need to handle the Settings item Cancel label updates.

majora2007 committed:
Changed the Cancel wording to Close after you edit a setting item.

majora2007 commented:
I've migrated all the preferences over and setup some new themeable button styles. 

Before tackling the series detail changes, I want to do a holistic run through the app and clean it up so we can push this out. I'm blocked on other issues due to the scope of these changes.

therobbiedavis committed:
positioning and css clean up initial pass

still lots to do

therobbiedavis committed:
more css updates

majora2007 commented on UI/Web/src/theme/themes/dark.scss:
```diff
@@ -33,11 +33,11 @@
     --primary-color-darkest-shade: #25624A;
     --error-color: #BD362F;
-    --bs-body-bg: #3f4245; // #121212
+    --bs-body-bg: #000; // #121212
```
Robbie we can't have this changed. It breaks everything:
![image](https://github.com/user-attachments/assets/67d392c0-4d70-45f3-912e-555af8bd61a1)

therobbiedavis replied:
Reverted already

majora2007 committed:
Updated backend dependencies

majora2007 committed:
Updated active event counter to use a debounced observable to avoid flashing of when activity is going on with event widget.

majora2007 committed:
Reverted Robbie's black background change.

Moved the customize from Home actionables to the settings menu which has much more space.

majora2007 committed:
Generic tweaks to the styling on the customize pages.

majora2007 committed:
Removed ngIf from all customize components

majora2007 committed:
Added a new button class called btn-primary-outline which is the primary button with transparent background.

majora2007 committed:
Tweaked the styles of btn-secondary-outline

therobbiedavis committed:
Merge branch 'feature/revamped-ux' of https://github.com/Kareadita/Kavita into feature/revamped-ux

therobbiedavis committed:
Changing some preference sizing

therobbiedavis committed:
login background, button, hover bg

majora2007 committed:
Tweaked the hover effect for btn-secondary-outline

majora2007 committed:
Recoded the import MAL Stack to a dedicated non-modal component.

majora2007 committed:
Recoded the import CBL to a dedicated non-modal component.

majora2007 committed:
Fixed bad background color on modals.

majora2007 committed:
Cleaned up edit user modal for some better spacing.

majora2007 committed:
Updated the styles in library-selector.

majora2007 committed:
Updated the styles in role-selector and added a select/deselect all checkbox to mimic style in library selector.

majora2007 committed:
Lots of cleanup around manage components

therobbiedavis committed:
Updating invite user flow

therobbiedavis committed:
Updating add library flow

majora2007 committed:
Merged develop in

majora2007 committed:
Moved the custom button location to be right aligned.

majora2007 committed:
A touch of paint across the app and fixed reading lists having non-nullable Primary/Secondary colors which broke unit tests. Started moving some code so we can prepare for colorscape support on readinglist/collections.

majora2007 committed:
Implemented a framework for the preference side nav so that when there are alerts, the user can know and take action.

majora2007 committed:
Scrobble Error count now shows on pref side nav

majora2007 committed:
Major QoL on manage Users table

majora2007 committed:
Fixed indeterminate checkboxes not showing the primary color.

majora2007 committed:
Converted manage libraries to the table design as well.

therobbiedavis committed:
Fixing some mobile side-nav issues

therobbiedavis committed:
more side-nav fixes

majora2007 committed:
Ensure that the currently selected preference side nav item is in view when loading a (fresh) page.

majora2007 committed:
Trying a new look out for buttons.

majora2007 committed:
Apply custom positioning for Add Device button

therobbiedavis committed:
fixing dead space on login page

therobbiedavis committed:
fixing up some styles on task settings

majora2007 committed:
Update Media Issues to have a date of the issue and removed Details and Extension since details is too technical and extension is in the file already.

majora2007 committed:
Updated carousel component to remove ngIf

majora2007 committed:
Fixed promote reading list not working from single card actionable.

majora2007 committed:
Use a modal to show preview images for theme manager instead of opening a new browser tab.

majora2007 committed:
Fixed missing pagination on reading list opds feeds.

majora2007 committed:
Setting items green text (the view value) is now clickable to switch into edit mode.

majora2007 committed:
Added a button to theme manager to allow resetting the form to allow uploading your own theme from file.

majora2007 committed:
Minor cleanup

majora2007 committed:
Reverted series detail actionable being moved.

majora2007 committed:
Added a dedicated task that generates colorscapes if they don't exist for a whole library.

majora2007 committed:
Removed Home side nav item on default side nav

majora2007 committed:
Fixed the search section not having a good background after tweaking other areas.

majora2007 committed:
Don't show empty sections on preference side nav

majora2007 committed:
Don't call apis for admins on shared pref sidenav items

majora2007 committed:
Fixed no-data on a table not being center aligned.

majora2007 committed:
Don't show the add button on theme manager when not an admin.

majora2007 committed:
Fixed an issue where api key was disappearing when clicking show/copy

therobbiedavis committed:
fixing pref side nav spacing

majora2007 committed:
Made announcements page viewable for all users, not just admins.

Updated the code so that isAdmin$ check is driven from accountService and I'm not repeating logic everywhere.

therobbiedavis committed:
Fixing theme manager unused height

majora2007 committed:
Started a crazy idea

majora2007 committed:
You can now click off the setting item to close it automatically

therobbiedavis committed:
Fixing theme manager on mobile

majora2007 committed:
You can now click off the setting item to close it automatically

majora2007 committed:
Brought back the ability to collapse pref side nav

majora2007 committed:
Fixed a bad highlight color and fixed 5 missing localization areas.

majora2007 committed:
Fixed an edge case where busting Kavita+ cache in edit series then hitting close wasn't refreshing screen.

therobbiedavis committed:
Fixing search and background canvas

therobbiedavis committed:
Merge branch 'feature/revamped-ux' of https://github.com/Kareadita/Kavita into feature/revamped-ux

majora2007 committed:
Fixed event widget not showing.

majora2007 committed:
Fixed table not showing sort icons

majora2007 committed:
Fix a weird bootstrap style for switches.

majora2007 committed:
Implemented the ability to collapse buttons on manage library page down into actionabl
